contracts/src/PDPService.sol

```diff
@@ -253,31 +284,36 @@ contract PDPService {
         return rootId;
     }
 
+    function scheduleRemovals(uint256 setId, uint256[] calldata rootIds) public {
+        require(rootIds.length + scheduledRemovals[setId].length <= MAX_ENQUEUED_REMOVALS, "Too many removals wait for next proving period to schedule");
+        require(proofSetOwner[setId] == msg.sender, "Only the owner can schedule removal of roots");
+        require(proofSetLive(setId), "Proof set not live");
+        uint256 totalDelta = 0;
+        for (uint256 i = 0; i < rootIds.length; i++){
+            totalDelta += rootLeafCounts[setId][rootIds[i]];
+            scheduledRemovals[setId].push(rootIds[i]);
```

Comment:
A removal of the _next_ root id followed by an add that takes that root id would result in the (delayed) removal trashing the subsequent add. And as bonus returning the wrong delta here (it can't know).

I think this needs to check that the scheduled-to-remove IDs are <= the next root ID at least. There is still scope for this return value to be misleading. For example, repeatedly removing the same root. Perhaps duplicates should also be caught here (by iterating the array?).

Please add tests for these cases.

Sending the notification only when actually removing the root could simplify things (but affects the current ability of the listener to abort and prevent the removal)

Reply:
Great points

I was back and forth between checking if root has been added before deleting.  I think it is better to do that check.

The more I think about it the total delta was just a guess about what the user wanted.  But the listener has all this information from addRoots. I'm more inclined to allow duplicates in the queue and drop any information about the delta.  This then makes removal match add better.

Reply:
Dropping delta SGTM